pkg/batcher/transport_message_batcher.go

```diff
@@ -0,0 +1,189 @@
+ * limitations under the License.
+ */
+
+package batcher
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/th2-net/th2-common-go/pkg/queue/message"
+	transport "github.com/th2-net/transport-go/pkg"
+)
+
+type Protocol = string
+
+const (
+	th2TransportProtocolAttribute = "transport-group"
+)
+
+type MessageArguments struct {
+	Metadata  map[string]string
+	Alias     string
+	Direction Direction
+	Protocol  Protocol
+}
+
+type messageBatcher struct {
+	encoder    transport.Encoder
+	groupIndex int
+
+	pipe        chan transport.RawMessage
+	seqProvider sequenceProvider
+
+	batchSizeBytes int
+	flushTimeout   time.Duration
+
+	router   message.Router
+	book     string
+	group    string
+	protocol Protocol
+
+	done chan bool
+}
+
+func NewMessageBatcher(router message.Router, cfg MqMessageBatcherConfig) (MqBatcher[MessageArguments], error) {
+	flushTimeout := cfg.FlushMillis
+	if flushTimeout <= 0 {
+		flushTimeout = DefaultFlushTime
+	}
+
+	maxBatchSizeBytes := cfg.BatchSizeBytes
+	if maxBatchSizeBytes <= 0 {
+		maxBatchSizeBytes = DefaultBatchSize
+	}
+
+	channelSize := cfg.ChannelSize
+	if channelSize <= 0 {
+		channelSize = DefaultChanelSize
+	}
+
+	flushDuration := time.Duration(flushTimeout) * time.Millisecond
+
+	batcher := messageBatcher{
+		encoder:        transport.NewEncoder(make([]byte, maxBatchSizeBytes)),
+		pipe:           make(chan transport.RawMessage, channelSize),
+		done:           make(chan bool, 1),
+		seqProvider:    sequenceProvider{},
+		batchSizeBytes: int(maxBatchSizeBytes),
+		flushTimeout:   flushDuration,
+		router:         router,
+		protocol:       cfg.Protocol,
+		group:          cfg.Group,
+		book:           cfg.Book,
+	}
+	go batcher.flushingRoutine()
+	return &batcher, nil
+}
+
+func (b *messageBatcher) Send(data []byte, args MessageArguments) error {
+	dataLength := len(data)
+	if dataLength <= 0 {
+		return nil
+	}
+
+	if dataLength > int(b.batchSizeBytes) {
+		return fmt.Errorf("too large message of data %d, max %d", dataLength, b.batchSizeBytes)
+
+	}
+
+	b.pipe <- transport.RawMessage{
+		MessageId: transport.MessageId{
+			SessionAlias: args.Alias,
+			Direction:    args.Direction,
+		},
+		Protocol: orDefaultIfEmpty(args.Protocol, b.protocol),
+		Metadata: args.Metadata,
+		Body:     data,
+	}
+
+	return nil
+}
+
+func (b *messageBatcher) Close() error {
+	log.Info().Msg("Closing message batcher")
+	close(b.pipe)
+	<-b.done
+	return nil
+}
+
+func (b *messageBatcher) flushingRoutine() {
+	log.Info().Msg("Flushing routine started")
+	defer func() {
+		b.done <- true
+	}()
+	for {
+		timer := time.After(b.flushTimeout)
+
+		select {
+		case item, ok := <-b.pipe:
+			if !ok {
+				log.Debug().Msg("Flushing messages by pipe complete")
+				b.flush()
+
+				log.Info().Msg("Flushing routine stopped")
+				return
+			}
+			newSize := b.encoder.SizeAfterEncodeRaw(b.group, b.book, item, b.groupIndex)
+			log.Trace().Msgf("Size after encoding: %d", newSize)
```

Comment:
note: please, use methods that add parameters to log message without allocating a new string

```suggestion
			log.Trace().Int("newSize", newSize).Int("currentSize", b.batchSizeBytes).Msg("approximate batch size computed")
```